### PR TITLE
9731 forward user

### PR DESCRIPTION
**Pull request by maimorag**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: [link to the issue](https://jira-dc.paloaltonetworks.com/browse/CIAC-9731)

## Description
- Added the functionality **API Key** and **API Key ID** (type 9) to Investigation & Response.
- Added the functionality **API Key** and **API Key ID** (type 9) toCortex Core - IOC.
- Added _http_request using **demisto.call** to IRApiModule.

### CODE REVIEW

File: Packs/ApiModules/pack_metadata.json [hunk 1]
```diff
@@ -2,7 +2,7 @@
     "name": "ApiModules",
     "description": "API Modules",
     "support": "xsoar",
-    "currentVersion": "2.2.22",
+    "currentVersion": "2.2.23",
```
Comment (dantavori) [inline Packs/ApiModules/pack_metadata.json:5]: no need to bump

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py [hunk 2]
```diff
@@ -148,6 +148,93 @@ class CoreClient(BaseClient):
...
+                    if status falls in ``status_forcelist`` range and retries have
+                    been exhausted.
+            
+            :type timeout: ``float`` or ``tuple``
+            :param timeout:
+                The amount of time (in seconds) that a request will wait for a client to
+                establish a connection to a remote machine before a timeout occurs.
+                can be only float (Connection Timeout) or a tuple (Connection Timeout, Read Timeout).
+        '''
+        if not is_demisto_version_ge(version='8.6.0',build_number='924770'):
```
Comment (dantavori) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:191]:
```suggestion
        if not is_xsiam() or not is_demisto_version_ge(version='8.6.0',build_number='924770'):
```

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py [hunk 3]
```diff
@@ -147,27 +147,8 @@ def main():  # pragma: no cover
     args = demisto.args()
     args["integration_context_brand"] = INTEGRATION_CONTEXT_BRAND
     args["integration_name"] = INTEGRATION_NAME
-    api_key = demisto.params().get('apikey')
-    api_key_id = demisto.params().get('apikey_id')
-    url = demisto.params().get('url')
+    url = "/api/webapp/"
```
Comment (dantavori) [inline Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py:150]: what if the build number is lower than 924770?

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py [hunk 4]
```diff
@@ -148,6 +148,93 @@ class CoreClient(BaseClient):
...
+            return BaseClient._http_request(self, method=method, url_suffix=url_suffix, full_url=full_url, headers=None,
+                                            json_data=json_data,params=params, data=data, timeout=timeout,
+                                            raise_on_status=raise_on_status)
+        try:
+            # Replace params if supplied
+            address = full_url if full_url else urljoin(self._base_url, url_suffix)
```
Comment (dantavori) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:197]: does demisto._apiCall support full_url?

**Review by dantavori — CHANGES_REQUESTED** [review]
nice, make sure to test the changes also in XSIAM with build number lower than the specified

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule_test.py [hunk 5]
```diff
@@ -3908,3 +3908,27 @@ def test_xsoar_to_xdr_flexible_close_reason_mapping(capfd, mocker, custom_mappin
 
         assert remote_args.delta.get('status')
         assert remote_args.delta['status'] == expected_resolved_status[i]
+
```
Comment (dantavori) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule_test.py:3911]: please add a test for the expected behavior for each product type and build

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py (hunk 3 above)
Comment (maimorag) [reply Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py:150]: i'll add the check here as well

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py [hunk 6]
```diff
@@ -148,6 +148,93 @@ class CoreClient(BaseClient):
...
+                        method=method,
+                        path=address,
+                        data=data,
+                        headers=self._headers,
+                        timeout=timeout
+                    )
+            if not self._is_status_code_valid(res, ok_codes):
+                    self._handle_error(error_handler, res, with_metrics)
+            return json.loads(res.get('data'))
+        except requests.exceptions.ConnectTimeout as exception:
```
[the remark this answers is not in the document]
Comment (maimorag) [reply Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:210]: According to Ben, it should return the same errors as before

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py (hunk 4 above)
Comment (maimorag) [reply Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:197]: we can also pass full url and also suffix to address

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule_test.py [hunk 7]
```diff
@@ -3915,3 +3915,52 @@ def test_xsoar_to_xdr_flexible_close_reason_mapping(capfd, mocker, custom_mappin
...
+    Then
+        - make sure the correct http_request is being called.
+    """
+    from CoreIRApiModule import CoreClient
+    import CoreIRApiModule
+    client = CoreClient(
+            base_url=f'{Core_URL}/public_api/v1', headers={}
+        )
+    mocker.patch.object(CoreIRApiModule, 'is_demisto_version_ge', return_value=True)
+    mocker.patch.object(CoreIRApiModule, 'is_xsiam', return_value=True)
```
Comment (dantavori) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule_test.py:3936]:
```suggestion
    mocker.patch.object(demisto, 'demistoVersion', return_value={"platform": "x2", "version": "8.6.0", "buildNumber": "957366"})
```

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule_test.py [hunk 8]
```diff
@@ -3915,3 +3915,52 @@ def test_xsoar_to_xdr_flexible_close_reason_mapping(capfd, mocker, custom_mappin
...
+    
+    mocker.patch.object(demisto, "_apiCall" ,return_value={'name': '/api/webapp/public_api/v1/distributions/get_versions/',
+                                         'status': 200,
+                                         'data': '{"reply":[{"container": ["1.1.1.1"]}]'})
+    res = client._http_request(method = "POST",
+                                url_suffix="/distributions/get_versions/")
+    assert res['data'] == {"reply":[{"container": ["1.1.1.1"]}]}
+
+
+def test_http_request_base_client(mocker):
```
Comment (dantavori) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule_test.py:3946]: let's do a similar change here as suggested above

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py [hunk 9]
```diff
@@ -147,27 +147,31 @@ def main():  # pragma: no cover
...
-
-    if not api_key or not api_key_id or not url:
-        headers = {
-            "HOST": demisto.getLicenseCustomField("Core.ApiHostName"),
-            demisto.getLicenseCustomField("Core.ApiHeader"): demisto.getLicenseCustomField("Core.ApiKey"),
-            "Content-Type": "application/json"
-        }
-        url = "http://" + demisto.getLicenseCustomField("Core.ApiHost") + "/api/webapp/"
-        add_sensitive_log_strs(demisto.getLicenseCustomField("Core.ApiKey"))
+    if not is_xsiam() or not is_demisto_version_ge(version='8.6.0',build_number='924770'):
```
Comment (dantavori) [inline Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py:152]: why different build number?

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py [hunk 10]
```diff
@@ -177,12 +181,11 @@ def main():  # pragma: no cover
     except ValueError as e:
         demisto.debug(f'Failed casting timeout parameter to int, falling back to 120 - {e}')
         timeout = 120
-
     client = Client(
         base_url=base_url,
         proxy=proxy,
         verify=verify_cert,
-        headers=headers,
+        headers={},
```
Comment (dantavori) [inline Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py:188]: you are not using the headers defined in lines 158-171

**Review by dantavori — CHANGES_REQUESTED** [review]
nice, see notes

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py [hunk 11]
```diff
@@ -142,14 +142,107 @@
     "resourceType",
     "tenantId",
 }
+RBAC_VALIDATIONS_VERSION = '8.6.0'
+RBAC_VALIDATIONS_BUILD_NUMBER = '957366'
 
 
 class CoreClient(BaseClient):
+    forward_user_run_RBAC_validation = is_xsiam() and is_demisto_version_ge(version=RBAC_VALIDATIONS_VERSION,
```
[the remark this answers is not in the document]
Comment (maimorag) [reply Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:150]: ok, I changed it to a global variable, thanks!
Should I remove the build check @dantavori?

File: Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py (hunk 9 above)
Comment (maimorag) [reply Packs/Core/Integrations/CortexCoreIR/CortexCoreIR.py:152]: fixed by using the global param

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py (hunk 4 above)
Comment (maimorag) [reply Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:204]: we don't, thanks

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py [hunk 12]
```diff
@@ -142,14 +142,82 @@
...
+                        path=address,
+                        data=data,
+                        headers=self._headers,
+                        timeout=timeout
+                    )
+            if ok_codes and response.get('status') not in ok_codes:
+                    self._handle_error(error_handler, response, with_metrics)
+            return json.loads(response['data'])
+        except Exception as exception:
+            raise DemistoException(exception)
```
Comment (dantavori) [inline Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py:218]: I'd remove the try-except, now you just replace the exception type to DemistoException, better keep the original

File: Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py (hunk 4 above)
Comment (dantavori) [reply Packs/ApiModules/Scripts/CoreIRApiModule/CoreIRApiModule.py]: ok, was it tested?

File: Packs/Core/ReleaseNotes/3_0_29.md [hunk 13]
```diff
@@ -0,0 +1,11 @@
+#### Integrations
+
+##### Investigation & Response
+
+- Improved implementation of the authorization process.
+- Updated the Docker image to: *demisto/python3:3.10.14.91134*.
+
+##### Cortex Core - IOC
+
+- Improved implementation of the authorization process
```
[the remark this answers is not in the document]
Comment (dantavori) [reply Packs/Core/ReleaseNotes/3_0_29.md:10]: why resolved without handling?

File: Packs/Core/Integrations/CoreIOCs/CoreIOCs_test.py [hunk 14]
```diff
@@ -80,8 +80,9 @@ def test_with_json(self):
             Then:
                 - the json ready to send
         """
+        client = Client(params={})
         _json = {'test': 'test'}
-        output = get_requests_kwargs(_json=_json)
+        output = get_requests_kwargs(client, _json=_json)
```
Comment (dantavori) [inline Packs/Core/Integrations/CoreIOCs/CoreIOCs_test.py:85]:
```suggestion
        output = get_requests_kwargs(_json=_json)
```

**Review by dantavori — COMMENTED** [review]
overall looks good, see last notes. I assume we need to test it again E2E across all platforms/versions before merging.

File: Packs/Core/ReleaseNotes/3_0_29.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (maimorag) [reply Packs/Core/ReleaseNotes/3_0_29.md]: I thought i merged this one, sorry

**Pull request closed without merging**